dynatrace_event/stackstate_checks/dynatrace_event/dynatrace_event.py

```diff
@@ -0,0 +1,264 @@
+        if not from_time:
+            from_time = int((datetime.now() - timedelta(days=self.events_boostrap_days)).strftime('%s'))*1000
+        # to_time = int(datetime.now().strftime('%s'))*1000
+        params = "?from={0}".format(from_time)
+        if cursor:
+            params = "?cursor={}".format(cursor)
+            endpoint = self.url + "/api/v1/events{}".format(params)
+        else:
+            endpoint = self.url + "/api/v1/events{}".format(params)
+        events = self.get_json_response(endpoint)
+        # check if the event limit exceeded condition reached
+        if self.event_limit_exceeded_condition(events):
+            return
+        if "error" not in events:
+            for item in events["events"]:
+                # if the limit reached for event_process_limit then break from loop
+                if self.should_stop_processing(events_processed):
+                    break
+                entityId = item.get("entityId")
+                event_status = item.get("eventStatus")
+                start_time = int(item.get("startTime"))
+                event_type = item.get("eventType")
+                # Create an event when no event exist for that entityId
+                state_events = self.status.data.get(self.url+"/events", {})
+                if self.url + "/events" not in self.status.data:
+                    self.status.data[self.url+"/events"] = {}
+                entity_events = state_events.get(entityId)
+                if entity_events is None:
+                    self.log.info("Creating a new open event type for an existing entityId")
+                    self.status.data[self.url + "/events"][entityId] = {event_type: item}
+                # Check if event already exist for that entityId
+                elif entity_events:
+                    event = entity_events.get(event_type)
+                    end_time = entity_events.get(event_type, {}).get("endTime")
+                    status = entity_events.get(event_type, {}).get("eventStatus")
+                    # different type of event can exist for same entityId
+                    if event is None and event_status == "OPEN":
+                        self.log.info("Creating a new open event type for an existing entityId")
+                        self.status.data[self.url + "/events"][entityId][event_type] = item
+                    # new event for same entityId with latest time
+                    elif start_time > end_time and status == event_status:
+                        self.log.info("Updating the existing event type for an existing entityId with new event")
+                        del self.status.data[self.url + "/events"][entityId][event_type]
+                        self.status.data[self.url + "/events"][entityId][event_type] = item
+                    # new event with CLOSED status come in for existing OPEN event for an even type of that entityId
+                    elif status == "CLOSED" and event_status == "OPEN":
```

Comment:
maybe write a comment that it is possible for a `CLOSED` event to reach the agent for which it never got an `OPEN` event and that in that case we simply ignore this event.

Reply:
discussed and agreed as not valid because we need to process since we go in past to check the event need to be removed or not.